File: android/app/src/main/java/com/esscey/makevibecodearcade/plugins/MakeCodeBridgePlugin.java
package com.esscey.makevibecodearcade.plugins;

import android.annotation.SuppressLint;
import android.view.ViewGroup;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@CapacitorPlugin(name = "MakeCodeBridge")
public class MakeCodeBridgePlugin extends Plugin {

    private WebView webView;
    private final ConcurrentHashMap<String, PluginCall> pendingCalls = new ConcurrentHashMap<>();

    // ── show ─────────────────────────────────────────────────────────────────

    @PluginMethod
    @SuppressLint("SetJavaScriptEnabled")
    public void show(PluginCall call) {
        String url = call.getString("url", "https://arcade.makecode.com");

        getActivity().runOnUiThread(() -> {
            if (webView != null) {
                webView.loadUrl(url);
                call.resolve();
                return;
            }

            WebView wv = new WebView(getContext());
            WebSettings settings = wv.getSettings();
            settings.setJavaScriptEnabled(true);
            settings.setDomStorageEnabled(true);
            settings.setAllowContentAccess(true);
            settings.setAllowFileAccess(true);
            settings.setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
            settings.setMediaPlaybackRequiresUserGesture(false);
            settings.setUseWideViewPort(true);
            settings.setLoadWithOverviewMode(true);

            wv.setWebViewClient(new WebViewClient());
            wv.setWebChromeClient(new WebChromeClient());
            wv.addJavascriptInterface(new BridgeInterface(), "MakeCodeAndroidBridge");

            // Insert behind the Capacitor WebView
            ViewGroup rootView = (ViewGroup) getBridge().getWebView().getParent();
            if (rootView == null) {
                call.reject("Cannot find root layout");
                return;
            }

            float density = getContext().getResources().getDisplayMetrics().density;
            int screenHeight = getContext().getResources().getDisplayMetrics().heightPixels;
            int tabBarPx = (int) (56 * density);

            ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                screenHeight - tabBarPx
            );
            rootView.addView(wv, 0, params);
            webView = wv;
            wv.loadUrl(url);
            call.resolve();
        });
    }

    // ── hide ─────────────────────────────────────────────────────────────────

    @PluginMethod
    public void hide(PluginCall call) {
        getActivity().runOnUiThread(() -> {
            if (webView != null) {
                ViewGroup parent = (ViewGroup) webView.getParent();
                if (parent != null) parent.removeView(webView);
                webView = null;
            }
            call.resolve();
        });
    }

    // ── executeScript ─────────────────────────────────────────────────────────

    @PluginMethod
    public void executeScript(PluginCall call) {
        String script = call.getString("script");
        if (script == null) {
            call.reject("Missing 'script' parameter");
            return;
        }
        if (webView == null) {
            call.reject("WebView not initialised — call show() first");
            return;
        }

        String callId = UUID.randomUUID().toString();
        pendingCalls.put(callId, call);

        String wrapped =
            "(async function() {\n" +
            "  try {\n" +
            "    const _result = await (async function() {\n" +
            script + "\n" +
            "    })();\n" +
            "    window.MakeCodeAndroidBridge.postResult('" + callId + "', JSON.stringify(_result != null ? _result : null), null);\n" +
            "  } catch (e) {\n" +
            "    window.MakeCodeAndroidBridge.postResult('" + callId + "', null, String(e.message || e));\n" +
            "  }\n" +
            "})();";

        getActivity().runOnUiThread(() -> webView.evaluateJavascript(wrapped, null));
    }

    // ── JavascriptInterface ───────────────────────────────────────────────────

    public class BridgeInterface {
        @JavascriptInterface
        public void postResult(String callId, String value, String error) {
            PluginCall call = pendingCalls.remove(callId);
            if (call == null) return;
            if (error != null) {
                call.reject(error);
            } else {
                JSObject result = new JSObject();
                result.put("result", value != null ? value : "null");
                call.resolve(result);
            }
        }
    }
}

File: native-plugins/android/MakeCodeBridgePlugin.kt
package com.pghnetworks.makevibecodearcade.plugins

import android.annotation.SuppressLint
import android.view.ViewGroup
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@CapacitorPlugin(name = "MakeCodeBridge")
class MakeCodeBridgePlugin : Plugin() {

    private var webView: WebView? = null
    private val pendingCalls = ConcurrentHashMap<String, PluginCall>()

    // MARK: - show

    @PluginMethod
    @SuppressLint("SetJavaScriptEnabled")
    fun show(call: PluginCall) {
        val url = call.getString("url") ?: "https://arcade.makecode.com"

        activity.runOnUiThread {
            if (webView != null) {
                webView!!.loadUrl(url)
                call.resolve()
                return@runOnUiThread
            }

            val wv = WebView(context).apply {
                settings.apply {
                    javaScriptEnabled = true
                    domStorageEnabled = true
                    allowContentAccess = true
                    allowFileAccess = true
                    mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
                    mediaPlaybackRequiresUserGesture = false
                    useWideViewPort = true
                    loadWithOverviewMode = true
                }
                webViewClient = WebViewClient()
                webChromeClient = WebChromeClient()
                addJavascriptInterface(BridgeInterface(), "MakeCodeAndroidBridge")
            }

            // Insert behind the Capacitor WebView, full-width with room for the tab bar
            val rootView = bridge.webView.parent as? ViewGroup ?: run {
                call.reject("Cannot find root layout")
                return@runOnUiThread
            }

            val displayMetrics = context.resources.displayMetrics
            val screenHeight = displayMetrics.heightPixels
            val tabBarDp = 56
            val tabBarPx = (tabBarDp * displayMetrics.density).toInt()
            val params = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                screenHeight - tabBarPx
            )
            rootView.addView(wv, 0, params)
            webView = wv
            wv.loadUrl(url)
            call.resolve()
        }
    }

    // MARK: - hide

    @PluginMethod
    fun hide(call: PluginCall) {
        activity.runOnUiThread {
            webView?.let { wv ->
                (wv.parent as? ViewGroup)?.removeView(wv)
                webView = null
            }
            call.resolve()
        }
    }

    // MARK: - executeScript

    @PluginMethod
    fun executeScript(call: PluginCall) {
        val script = call.getString("script") ?: run {
            call.reject("Missing 'script' parameter")
            return
        }
        val wv = webView ?: run {
            call.reject("WebView not initialised — call show() first")
            return
        }

        val callId = UUID.randomUUID().toString()
        pendingCalls[callId] = call

        val escaped = script.replace("\\", "\\\\").replace("`", "\\`")
        val wrapped = """
            (async function() {
              try {
                const _result = await (async function() {
                  $script
                })();
                window.MakeCodeAndroidBridge.postResult(
                  "$callId",
                  JSON.stringify(_result ?? null),
                  null
                );
              } catch (e) {
                window.MakeCodeAndroidBridge.postResult(
                  "$callId",
                  null,
                  String(e.message || e)
                );
              }
            })();
        """.trimIndent()

        activity.runOnUiThread {
            wv.evaluateJavascript(wrapped, null)
        }
    }

    // MARK: - JavascriptInterface for result callbacks

    inner class BridgeInterface {
        @JavascriptInterface
        fun postResult(callId: String, value: String?, error: String?) {
            val call = pendingCalls.remove(callId) ?: return
            if (error != null) {
                call.reject(error)
            } else {
                val result = JSObject()
                result.put("result", value ?: "null")
                call.resolve(result)
            }
        }
    }
}
